File: python/cleaners.py
import re

def clean_double_newline_markdown_links(text):
    """
    Remove double newlines with backslashes in markdown links within a file

    This function cleans up markdown-style links that contain double newlines
    with backslashes, while preserving emojis and the overall link structure.
    It processes the entire file, only modifying the specific markdown links.

    >>> text = "Some text\\n[📄️ Metrics\\\\\\n\\\\\\nTo monitor...](/docs/metrics/)\\nMore text"
    >>> clean_double_newline_markdown_links(text)
    'Some text\\n[📄️ Metrics To monitor...](/docs/metrics/)\\nMore text'
    """
    def replace_link(match):
        full_content = match.group(1)
        url = match.group(2)
        
        cleaned_content = re.sub(r'\\\s*\n\s*\\\s*\n\s*', ' ', full_content)
      
        return f'[{cleaned_content}]({url})'

    pattern = r'\[(.*?\\\s*\n\s*\\\s*\n\s*.*?)\]\((.*?)\)'
    cleaned_text = re.sub(pattern, replace_link, text, flags=re.DOTALL)

    return cleaned_text


def clean_anchortag_headings(text):
    """
    Replace anchortag headings with h2 tags in markdown text.

    This function replaces markdown-style anchortag headings with h2 tags.
    It processes the entire text, modifying only the specific heading patterns.

    >>> text = "Some text\\n[](#heading1)\\nHeading 1\\nMore text\\n[](#heading2)\\nHeading 2"
    >>> clean_anchortag_headings(text)
    'Some text\\n## Heading 1\\nMore text\\n## Heading 2'
    """
    return re.sub(r'\[\]\((#.*?)\)\n(.*?)(?=\n|$)', r'## \2', text)


def remove_end_matter(text):
    """
    Remove end matter including "[](#get-help)", "[Prev", and specific help text.

    This function finds the earliest occurrence of "[](#get-help)", "[Prev",
    or a specific help text, and removes everything from that point onwards.

    >>> text = "Some content\\n[](#get-help)\\nGet help here\\n[Prev]\\nPrevious content"
    >>> remove_end_matter(text)
    'Some content'
    >>> text = "Main text\\nIf you have any questions or need any help in setting things up, join our slack community and ping us in `#help` channel.\\nMore text"
    >>> remove_end_matter(text)
    'Main text'
    >>> text = "Only main content even if help mentioned"
    >>> remove_end_matter(text)
    'Only main content even if help mentioned'
    """
    remove_index = min(
        (text.find("[](#get-help)"),
         text.find("[Prev"),
         text.find("If you have any questions or need any help in setting things up, join our slack community and ping us in `#help` channel.")),
        key=lambda x: float('inf') if x == -1 else x
    )
    if remove_index != -1:
        return text[:remove_index].strip()
    return text





if __name__ == "__main__":
    import doctest
    doctest.testmod(verbose=True)

File: node/cleaners.js
const cleanDoubleNewlineMarkdownLinks = (text) => {
    /**
     * Remove double newlines with backslashes in markdown links within a file
     *
     * This function cleans up markdown-style links that contain double newlines
     * with backslashes, while preserving emojis and the overall link structure.
     * It processes the entire file, only modifying the specific markdown links.
     *
     * @param {string} text - The input text to clean
     * @returns {string} The cleaned text
     */
    const replaceLink = (match, fullContent, url) => {
        const cleanedContent = fullContent.replace(/\\\s*\n\s*\\\s*\n\s*/g, ' ');
        return `[${cleanedContent}](${url})`;
    };

    const pattern = /\[(.*?\\\s*\n\s*\\\s*\n\s*.*?)\]\((.*?)\)/gs;
    return text.replace(pattern, replaceLink);
};

const cleanAnchortagHeadings = (text) => {
    /**
     * Replace anchortag headings with h2 tags in markdown text.
     *
     * This function replaces markdown-style anchortag headings with h2 tags.
     * It processes the entire text, modifying only the specific heading patterns.
     *
     * @param {string} text - The input text to clean
     * @returns {string} The cleaned text
     */
    return text.replace(/\[\]\((#.*?)\)\n(.*?)(?=\n|$)/g, '## $2');
};

const cleanDoubleAsteriskWhitespaceGaps = (text) => {
    /**
     * When double asterisks appear immediately before the start of a link,
     * they may appear after the link after a newline.
     * 
     * Example:
     * ```
     * **[Use the all-in-one auto-instrumentation library(Recommended)](#using-the-all-in-one-auto-instrumentation-library)\n    **
     * ```
     *
     * This function processes the entire text, replacing the whitespace before the second double asterisk pair
     * with an empty string.
     *
     * @param {string} text - The input text to clean
     * @returns {string} The cleaned text
     */
    return text.replace(/\*\*(\[.*\]\(.*\))\n\s*\*\*/g, '**$1**');
};


const cleanNewlineAndSpacesAfterLinks = (text) => {
    /**
     * Remove newline after links followed by lowercase characters in markdown.
     *
     * This function removes a newline and trims spaces to one when a markdown
     * link is immediately followed by a newline, arbitrary spaces, and then
     * a lowercase character.
     *
     * @param {string} text - The input text to clean
     * @returns {string} The cleaned text
     */
    // This regex replaces a pattern where a markdown link is followed by a newline,
    // optional whitespace, and then a lowercase letter.
    // The replacement keeps the link ($1), adds a single space,
    // and preserves the content that follows ($2).
    // This ensures we keep the subsequent text "alive" in the output.
    return text.replace(/(\[.*?\]\(.*?\))\n\s*([a-z].*)/g, '$1 $2');
};



const cleanMultiColumnLinks = (markdownText) => {
    /**
     * Spread multi-column links in markdown text into a list format.
     *
     * This function takes markdown text containing paragraphs with multiple links
     * and transforms them into a list format, with each link on a new line.
     *
     * @param {string} markdownText - The input markdown text to clean
     * @returns {string} The cleaned markdown text
     */
    const linkPattern = /(\n\n)(\[(?:[^\]]+\\\s*)+[^\]]+\]\([^\)]+\)(?:\s*\[(?:[^\]]+\\\s*)+[^\]]+\]\([^\)]+\))*)\s*(?=$|\n\n)/g;
    
    const cleanLinks = (match, newlines, links) => {
        const cleanedLinks = links.match(/\[([^\]]+)\]\(([^\)]+)\)/g).map(link => {
            const [, linkText, linkUrl] = link.match(/\[([^\]]+)\]\(([^\)]+)\)/);
            const cleanText = linkText.replace(/\\\s*\n\s*\\\s*\n\s*/g, ': ')
                                      .replace(/\s*\\\s*\n\s*/g, ' ')
                                      .replace(/\\ \\ /g, ': ')
                                      .trim();
            return `- [${cleanText}](${linkUrl})`;
        });
        return newlines + cleanedLinks.join('\n');
    };

    return markdownText.replace(linkPattern, cleanLinks).trim();
};



const cleanExtraNewlinesAfterLinks = (text) => {
    /**
     * Remove extra newlines after links in markdown text.
     *
     * This function addresses two common cases:
     * 1. Removes a newline between a link and a period.
     * 2. Removes a newline immediately after a link when followed by a space.
     *
     * @param {string} text - The input text to clean
     * @returns {string} The cleaned text with extra newlines after links removed
     */
    // Remove newline between link and period
    text = text.replace(/(\[.*?\]\(.*?\))\n\./g, '$1.');
    
    // Remove newline immediately after link when it is followed by a space
    text = text.replace(/(\[.*?\]\(.*?\))\n (?=\S)/g, '$1 ');
    
    return text;
};


const removeEndMatter = (text) => {
    /**
     * Remove end matter including "[](#get-help)", "[Prev", and specific help text.
     *
     * This function finds the earliest occurrence of "[](#get-help)", "[Prev",
     * or a specific help text, and removes everything from that point onwards.
     *
     * @param {string} text - The input text to clean
     * @returns {string} The cleaned text
     */
    const patterns = [
        /\[]\(#get-help\)/,
        /\[Prev/,
        /If you have any questions or need any help in setting things up, join our slack community and ping us in `#help` channel./
    ];

    const indices = patterns.map(pattern => text.search(pattern))
                            .filter(index => index !== -1);

    if (indices.length > 0) {
        const removeIndex = Math.min(...indices);
        return text.slice(0, removeIndex).trim();
    }

    return text;
};

const cleaners = {
    cleanDoubleNewlineMarkdownLinks,
    cleanAnchortagHeadings,
    cleanExtraNewlinesAfterLinks,
    removeEndMatter,
    cleanMultiColumnLinks,
    cleanNewlineAndSpacesAfterLinks,
    cleanDoubleAsteriskWhitespaceGaps
};

export default cleaners;